thunx-pdp-opa/src/main/java/com/contentgrid/thunx/pdp/opa/QuerySetToThunkExpressionConverter.java

```diff
@@ -279,6 +287,89 @@ public ThunkExpression<?> visit(Null nullValue) {
 
         @Override
         public ThunkExpression<?> visit(ArrayTerm arrayTerm) {
+            ScalarCollectingRegoVisitor collectingRegoVisitor = new ScalarCollectingRegoVisitor();
+            arrayTerm.getValue().forEach(scalarTerm -> scalarTerm.accept(collectingRegoVisitor));
+            return Scalar.of(collectingRegoVisitor.getScalars());
+        }
+
+        @Override
+        public ThunkExpression<?> visit(SetTerm setTerm) {
+            ScalarCollectingRegoVisitor collectingRegoVisitor = new ScalarCollectingRegoVisitor();
+            setTerm.getValue().forEach(scalarTerm -> scalarTerm.accept(collectingRegoVisitor));
+            return Scalar.of(collectingRegoVisitor.getScalars());
+        }
+    }
+
+    static class ScalarCollectingRegoVisitor implements RegoVisitor<ThunkExpression<?>> {
+
+        @Getter
+        List<Scalar<?>> scalars = new ArrayList<>();
+
+        @Override
+        public ThunkExpression<?> visit(QuerySet querySet) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Query query) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Expression expression) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Ref ref) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Call call) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Var var) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Numeric numeric) {
+            Scalar<?> scalar = Scalar.of(numeric.getValue());
+            scalars.add(scalar);
+            return scalar;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Text text) {
+            Scalar<?> scalar = Scalar.of(text.getValue());
+            scalars.add(scalar);
+            return scalar;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Bool bool) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Null aNull) {
+            Scalar<?> scalar = Scalar.nullValue();
+            scalars.add(scalar);
+            return scalar;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(ArrayTerm arrayTerm) {
+            arrayTerm.getValue().forEach(scalarTerm -> scalarTerm.accept(this));
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(SetTerm setTerm) {
+            setTerm.getValue().forEach(scalarTerm -> scalarTerm.accept(this));
```

Comment:
Same remark as for the array case, nested sets shouldn't be flattened.

Reply:
fixed